File: packages/vav-ios/VAVRemote/VAVRemote/DiagLog.swift
import Foundation
import Network
import Tcmobile
import UIKit

/// Ring + file diagnostic log so a failed WAN dial can be exported from Settings.
enum DiagLog {
    private static let maxLines = 800
    private static let queue = DispatchQueue(label: "vav.remote.diag")
    private static var lines: [String] = []
    private static var monitor: NWPathMonitor?
    private static var lastPath = ""

    private static var fileURL: URL {
        FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask)[0]
            .appendingPathComponent("vav-remote-diag.log")
    }

    static func start() {
        queue.sync {
            if let existing = try? String(contentsOf: fileURL, encoding: .utf8), !existing.isEmpty {
                lines = existing.split(whereSeparator: \.isNewline).map(String.init)
                if lines.count > maxLines { lines = Array(lines.suffix(maxLines)) }
            }
        }
        line("diag start \(header())")
        startPathMonitor()
    }

    static func line(_ message: String) {
        let stamp = Self.stamp()
        let text = "\(stamp) \(redact(message))"
        queue.async {
            lines.append(text)
            if lines.count > maxLines { lines = Array(lines.suffix(maxLines)) }
            persistLocked()
        }
        NSLog("[vav-diag] %@", text)
    }

    static func ingestGo() {
        let go = TcmobileSnapshotLogs().trimmingCharacters(in: .whitespacesAndNewlines)
        guard !go.isEmpty else { return }
        queue.sync {
            for raw in go.split(whereSeparator: \.isNewline) {
                let row = String(raw)
                if !lines.contains(where: { $0.hasSuffix(row) || $0.contains(row) }) {
                    lines.append("\(stamp()) [tc] \(redact(row))")
                }
            }
            if lines.count > maxLines { lines = Array(lines.suffix(maxLines)) }
            persistLocked()
        }
    }

    static func snapshot() -> String {
        ingestGo()
        return queue.sync {
            ([header(), ""] + lines).joined(separator: "\n")
        }
    }

    static func exportFile() -> URL {
        let url = FileManager.default.temporaryDirectory
            .appendingPathComponent("vav-remote-diag.txt")
        try? snapshot().data(using: .utf8)?.write(to: url)
        return url
    }

    static func clear() {
        TcmobileClearLogs()
        queue.sync {
            lines = []
            persistLocked()
        }
        line("diag cleared")
    }

    private static func header() -> String {
        let device = UIDevice.current
        let info = Bundle.main.infoDictionary
        let version = info?["CFBundleShortVersionString"] as? String ?? "?"
        let build = info?["CFBundleVersion"] as? String ?? "?"
        return "VAV Remote \(version) (\(build)) iOS \(device.systemVersion) \(device.name) model=\(device.model)"
    }

    private static func startPathMonitor() {
        queue.async {
            guard monitor == nil else { return }
            let probe = NWPathMonitor()
            probe.pathUpdateHandler = { path in
                let summary = describe(path)
                queue.async {
                    guard summary != lastPath else { return }
                    lastPath = summary
                    line("path \(summary)")
                }
            }
            probe.start(queue: queue)
            monitor = probe
        }
    }

    private static func describe(_ path: NWPath) -> String {
        var kinds: [String] = []
        if path.usesInterfaceType(.wifi) { kinds.append("wifi") }
        if path.usesInterfaceType(.cellular) { kinds.append("cellular") }
        if path.usesInterfaceType(.wiredEthernet) { kinds.append("wired") }
        if path.usesInterfaceType(.other) { kinds.append("other") }
        if path.usesInterfaceType(.loopback) { kinds.append("loopback") }
        let status: String
        switch path.status {
        case .satisfied: status = "satisfied"
        case .unsatisfied: status = "unsatisfied"
        case .requiresConnection: status = "requiresConnection"
        @unknown default: status = "unknown"
        }
        return "\(status) via=\(kinds.joined(separator: "+")) expensive=\(path.isExpensive) constrained=\(path.isConstrained)"
    }

    private static func persistLocked() {
        try? lines.joined(separator: "\n").data(using: .utf8)?.write(to: fileURL)
    }

    private static func stamp() -> String {
        let f = DateFormatter()
        f.locale = Locale(identifier: "en_US_POSIX")
        f.dateFormat = "HH:mm:ss.SSS"
        return f.string(from: Date())
    }

    /// Drop pairing secrets / full tailcat tokens from anything we persist or share.
    static func redact(_ text: String) -> String {
        var out = text
        if let regex = try? NSRegularExpression(pattern: "tc[A-Za-z0-9_-]{10,}") {
            let range = NSRange(out.startIndex..., in: out)
            out = regex.stringByReplacingMatches(in: out, range: range, withTemplate: "tc…")
        }
        if let regex = try? NSRegularExpression(pattern: #""(secret|auth)"\s*:\s*"[^"]+""#) {
            let range = NSRange(out.startIndex..., in: out)
            out = regex.stringByReplacingMatches(in: out, range: range, withTemplate: "\"$1\":\"…\"")
        }
        return out
    }

    static func tokenHint(_ token: String) -> String {
        if token.count <= 8 { return "len=\(token.count)" }
        return "\(token.prefix(6))… len=\(token.count)"
    }
}

File: packages/vav-android/VAVRemote/app/src/main/java/com/vav/remote/DiagLog.kt
package com.vav.remote

import android.content.Context
import android.net.ConnectivityManager
import android.net.Network
import android.net.NetworkCapabilities
import android.os.Build
import android.util.Log
import java.io.File
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.concurrent.Executors

/**
 * Ring + file diagnostic log so a failed WAN dial can be exported from Settings.
 * Same surface as iOS `DiagLog`: redact `tc…` tokens and pairing secrets.
 */
object DiagLog {
    private const val MAX_LINES = 800
    private val lock = Any()
    private val io = Executors.newSingleThreadExecutor()
    private val lines = ArrayList<String>()
    private var app: Context? = null
    private var lastPath = ""
    private var pathCallback: ConnectivityManager.NetworkCallback? = null

    private fun file(): File? {
        val dir = app?.cacheDir ?: return null
        return File(dir, "vav-remote-diag.log")
    }

    fun start(context: Context) {
        val appContext = context.applicationContext
        synchronized(lock) {
            app = appContext
            val existing = file()?.takeIf { it.exists() }?.readText(Charsets.UTF_8).orEmpty()
            if (existing.isNotBlank()) {
                lines.clear()
                lines.addAll(existing.split('\n').filter { it.isNotEmpty() })
                if (lines.size > MAX_LINES) {
                    val keep = lines.takeLast(MAX_LINES)
                    lines.clear()
                    lines.addAll(keep)
                }
            }
        }
        line("diag start ${header()}")
        startPathMonitor(appContext)
    }

    fun line(message: String) {
        val text = "${stamp()} ${redact(message)}"
        io.execute {
            synchronized(lock) {
                lines.add(text)
                if (lines.size > MAX_LINES) {
                    val keep = lines.takeLast(MAX_LINES)
                    lines.clear()
                    lines.addAll(keep)
                }
                persistLocked()
            }
        }
        Log.i("vav-diag", text)
    }

    fun ingestGo() {
        val go = snapshotGoLogs().trim()
        if (go.isEmpty()) return
        synchronized(lock) {
            for (raw in go.split('\n')) {
                val row = raw.trim()
                if (row.isEmpty()) continue
                if (lines.none { it.endsWith(row) || it.contains(row) }) {
                    lines.add("${stamp()} [tc] ${redact(row)}")
                }
            }
            if (lines.size > MAX_LINES) {
                val keep = lines.takeLast(MAX_LINES)
                lines.clear()
                lines.addAll(keep)
            }
            persistLocked()
        }
    }

    fun snapshot(): String {
        ingestGo()
        return synchronized(lock) {
            (listOf(header(), "") + lines).joinToString("\n")
        }
    }

    fun clear() {
        clearGoLogs()
        synchronized(lock) {
            lines.clear()
            persistLocked()
        }
        line("diag cleared")
    }

    fun tokenHint(token: String): String {
        if (token.length <= 8) return "len=${token.length}"
        return "${token.take(6)}… len=${token.length}"
    }

    fun redact(text: String): String {
        var out = text
        out = TC_TOKEN.replace(out, "tc…")
        out = SECRET_JSON.replace(out, "\"$1\":\"…\"")
        return out
    }

    private fun header(): String {
        val ctx = app
        val version =
            try {
                ctx?.packageManager?.getPackageInfo(ctx.packageName, 0)?.versionName ?: "?"
            } catch (_: Exception) {
                "?"
            }
        return "VAV Remote $version Android ${Build.VERSION.RELEASE} ${Build.MODEL}"
    }

    private fun startPathMonitor(context: Context) {
        if (pathCallback != null) return
        val cm = context.getSystemService(Context.CONNECTIVITY_SERVICE) as? ConnectivityManager ?: return
        val callback =
            object : ConnectivityManager.NetworkCallback() {
                override fun onAvailable(network: Network) {
                    logPath(cm, network)
                }

                override fun onCapabilitiesChanged(network: Network, caps: NetworkCapabilities) {
                    logPath(cm, network, caps)
                }

                override fun onLost(network: Network) {
                    line("path unsatisfied via=lost")
                }
            }
        try {
            cm.registerDefaultNetworkCallback(callback)
            pathCallback = callback
            cm.activeNetwork?.let { logPath(cm, it) }
        } catch (err: Exception) {
            line("path monitor fail ${err.message}")
        }
    }

    private fun logPath(
        cm: ConnectivityManager,
        network: Network,
        caps: NetworkCapabilities? = cm.getNetworkCapabilities(network)
    ) {
        val summary = describe(caps)
        synchronized(lock) {
            if (summary == lastPath) return
            lastPath = summary
        }
        line("path $summary")
    }

    private fun describe(caps: NetworkCapabilities?): String {
        if (caps == null) return "unsatisfied via=none"
        val kinds = mutableListOf<String>()
        if (caps.hasTransport(NetworkCapabilities.TRANSPORT_WIFI)) kinds.add("wifi")
        if (caps.hasTransport(NetworkCapabilities.TRANSPORT_CELLULAR)) kinds.add("cellular")
        if (caps.hasTransport(NetworkCapabilities.TRANSPORT_ETHERNET)) kinds.add("wired")
        if (caps.hasTransport(NetworkCapabilities.TRANSPORT_VPN)) kinds.add("vpn")
        if (kinds.isEmpty()) kinds.add("other")
        val status = if (caps.hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)) "satisfied" else "unsatisfied"
        val expensive = !caps.hasCapability(NetworkCapabilities.NET_CAPABILITY_NOT_METERED)
        val constrained = !caps.hasCapability(NetworkCapabilities.NET_CAPABILITY_NOT_CONGESTED)
        return "$status via=${kinds.joinToString("+")} expensive=$expensive constrained=$constrained"
    }

    private fun persistLocked() {
        try {
            file()?.writeText(lines.joinToString("\n"), Charsets.UTF_8)
        } catch (_: Exception) {
        }
    }

    private fun stamp(): String {
        val f = SimpleDateFormat("HH:mm:ss.SSS", Locale.US)
        return f.format(Date())
    }

    private fun snapshotGoLogs(): String {
        return invokeTcmobile("snapshotLogs") ?: invokeTcmobile("SnapshotLogs") ?: ""
    }

    private fun clearGoLogs() {
        invokeTcmobile("clearLogs")
        invokeTcmobile("ClearLogs")
    }

    private fun invokeTcmobile(method: String): String? {
        return try {
            val clazz = Class.forName("tcmobile.Tcmobile")
            val found = clazz.methods.firstOrNull { it.name == method && it.parameterCount == 0 } ?: return null
            found.invoke(null)?.toString()
        } catch (_: Exception) {
            null
        }
    }

    private val TC_TOKEN = Regex("tc[A-Za-z0-9_-]{10,}")
    private val SECRET_JSON = Regex("\"(secret|auth)\"\\s*:\\s*\"[^\"]+\"")
}
